src/psyclone/tests/psyGen_test.py

```diff
@@ -4174,6 +4192,58 @@ def test_fparser2astprocessor_process_declarations_intent(f2008_parser):
         Symbol.Access.READWRITE
 
 
+def test_fparser2astprocessor_process_declarations_stmt_functions(
+        f2008_parser):
+    '''Test that process_declarations method handles statement functions
+    appropriately.
+    '''
+    from fparser.common.readfortran import FortranStringReader
+    from fparser.two.Fortran2003 import Specification_Part
+    fake_parent = KernelSchedule("dummy_schedule")
+    processor = Fparser2ASTProcessor()
+
+    # If 'a' is not declared it could be a statement function, which are
+    # unsupported and produce a NotImplementedError.
+    reader = FortranStringReader("a(x) = 1")
+    fparser2spec = Specification_Part(reader).content[0]
+    with pytest.raises(NotImplementedError) as error:
+        processor.process_declarations(fake_parent, [fparser2spec], [])
+    assert "Could not process '" in str(error.value)
+    assert "'. Statement Function declarations are not supported." \
+        in str(error.value)
+
+    # If 'a' is declared in the symbol table as an array, it is an array
+    # assignment which belongs in the execution part.
+    fake_parent.symbol_table.add(Symbol('a', 'real', shape=[None]))
+    fake_parent.symbol_table.add(Symbol('x', 'real', shape=[]))
+    processor.process_declarations(fake_parent, [fparser2spec], [])
+    assert len(fake_parent.children) == 1
+    array = fake_parent.children[0].children[0]
+    assert isinstance(array, Array)
+    assert array.name is "a"
+
+    # Test that it works with multi-dimensional arrays
+    fake_parent = KernelSchedule("dummy_schedule")
+    reader = FortranStringReader("b(x, y) = 1")
+    fparser2spec = Specification_Part(reader).content[0]
+    fake_parent.symbol_table.add(Symbol('b', 'real', shape=[None, None]))
+    fake_parent.symbol_table.add(Symbol('x', 'real', shape=[]))
```

Comment:
Yes, they should be integers.

This is not tested in the Symbol table, as the declaration of each symbol is independent.
It could be checked on the fparser2AASTProcessor Part_Ref handler, but we don't have enough infrastructure for it, for instance `A(i+ 4)`, we would need to know that this BinaryOperation evaluates to an integer expression. We can think about this in the future but it would be a big modification for this PR. Also if(EXPR1) EXPR2, EXPR1 evaluates to boolean or EXPR2 is a complete statement would be similar.

Reply:
Oh yes, that is rather more complicated than I thought :-) We could start with checks for the case where an array dimension is just a Reference but you're right, that's something for another Issue.